### PR TITLE
Part 1: Macro-ify AWS client usage

### LINKED ISSUE #861: Dissolve AWS Gateway

### PR DESCRIPTION
**Pull request by dfellis**
Part of #861

### CODE REVIEW

File: src/modules/0.0.10/aws_acm_import/index.ts [hunk 1]
```diff
@@ -1,10 +1,28 @@
...
+async function importCertificate(client: any, input: ImportCertificateCommandInput) {
+  const res = await client.importCertificate(input);
+  const arn = res.CertificateArn ?? '';
+  let certificates: string[] = [];
+  let i = 0;
+   // Wait for ~1min until imported cert is available
+  do {
+    const start = Date.now();
+    await new Promise(r => setTimeout(r, 2000));
+    while (Date.now() - start < 2000); // Sleep for 2s
```
Comment (aguillenv) [inline src/modules/0.0.10/aws_acm_import/index.ts:20]: oh you re copying here the fixed hot-loops issue

File: src/services/aws_macros.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  pageName?: string,
+  pageSize = 25
+): ((client: any) => Promise<any[]>) {
+  if (pageName) {
+    return async (client: any) => {
+      const vals = [];
+      const paginator = paginateFn({
+        client,
+        pageSize,
+      }, {});
```
Comment (aguillenv) [inline src/services/aws_macros.ts:67]: that `{}` should be also a variable input for this function. I think it is used somewhere to filter.

File: src/services/aws_macros.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      }
+      return vals;
+    };
+  }
+}
+
+export function crudBuilder(
+  methodName: string,
+  argMapper: (args: any[]) => any,
+  retFormatter?: (arg0: any) => any
```
Comment (aguillenv) [inline src/services/aws_macros.ts:93]: what does `ret` stands for here?

Comment (aguillenv) [reply src/services/aws_macros.ts:93]: oh a return like function?

**Review by aguillenv — COMMENTED** [review]
looking good 🚀

File: src/modules/0.0.10/aws_acm_import/index.ts (hunk 1 above)
Comment (dfellis) [reply src/modules/0.0.10/aws_acm_import/index.ts:20]: :facepalm:

File: src/services/aws_macros.ts (hunk 2 above)
Comment (dfellis) [reply src/services/aws_macros.ts:67]: I was only going to make it available *if* we have a usage that exploits it. If not, then we can do this and simplify the macro functions.

File: src/services/aws_macros.ts (hunk 3 above)
Comment (dfellis) [reply src/services/aws_macros.ts:93]: It formats the returned value, so you get a response from AWS, then do some minor transform to either get a sub-object or mutate in some way or the other.

File: src/modules/0.0.10/aws_acm_list/index.ts [hunk 4]
```diff
@@ -62,11 +88,12 @@ export const AwsAcmListModule: Module2 = new Module2({
         read: async (ctx: Context, id?: string) => {
           const client = await ctx.getAwsClient() as AWS;
           if (id) {
-            const rawCert = await client.getCertificate(id);
+            // const rawCert = await client.getCertificate(id);
```
[the remark this answers is not in the document]
Comment (dfellis) [reply src/modules/0.0.10/aws_acm_list/index.ts:94]:
```suggestion
```

**Review by aguillenv — APPROVED** [review]

**Pull request merged**